src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -583,6 +576,252 @@ private bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
+                    }
+                    finally
+                    {
+                        _connection.Transport.Input.AdvanceTo(consumed, examined);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.ServerDisconnectedWithError(_logger, ex);
+                closeException = ex;
+            }
+            finally
+            {
+                // Dispose the connection
+                await _connection.DisposeAsync();
+                _connection = null;
+
+                // Stop the timeout timer.
+                timeoutTimer?.Dispose();
+            }
+
+            if (closeException != null)
+            {
+                Log.ShutdownWithError(_logger, closeException);
+            }
+            else
+            {
+                Log.ShutdownConnection(_logger);
+            }
+
+            // Cancel any outstanding invocations
+            CancelOutstandingInvocations(closeException);
+
+            // Fire-and-forget the closed event
+            _ = Task.Run(() =>
+            {
+                try
+                {
+                    Log.InvokingClosedEventHandler(_logger);
+                    Closed?.Invoke(closeException);
+                }
+                catch (Exception ex)
+                {
+                    Log.ErrorDuringClosedEvent(_logger, ex);
+                }
+            });
+        }
+
+        private ValueTask<FlushResult> WriteAsync(byte[] payload) =>
+            _connection.Transport.Output.WriteAsync(payload);
+
+        private async Task StopAsyncCoreWithinLock()
+        {
+            if (_connection == null)
+            {
+                // No-op if we're already stopped.
+                return;
+            }
+
+            Log.Stopping(_logger);
+
+            // Complete our write pipe, which should cause everything to shut down
+            Log.CompletingTransportPipe(_logger);
+            _connection.Transport.Output.Complete();
+
+            // Wait ServerTimeout for the server or transport to shut down.
+            var delayCts = new CancellationTokenSource();
+            Log.WaitingForReceiveLoopToTerminate(_logger);
+            var trigger = await Task.WhenAny(_receiveTask, Task.Delay(ServerTimeout, delayCts.Token));
+            if (trigger == _receiveTask)
+            {
+                // Stop the timer.
+                delayCts.Cancel();
+            }
+            else
+            {
+                Log.TimedOutWaitingForReceiveLoop(_logger);
+                // We timed out waiting for the server/transport to shut down, cancel it, this will cause
+                // a TimeoutException indicating the server timeout elapsed to be signalled as
+                // the exception in the Closed event.
+                _connection.Transport.Input.CancelPendingRead();
+
+                Log.WaitingForCanceledReceiveLoop(_logger);
+                await _receiveTask;
```

Comment:
No, I don't think we should catch it. Since we cancel it by cancelling the pipe, it will throw `TimeoutException` (see the `if(result.IsCanceled)` clause in `ReceiveLoop`) and we want that exception to manifest to the user since the server/transport did time out.